src/widget/box_constraints.rs

```diff
@@ -0,0 +1,429 @@
+                    height: self.min.width * aspect_ratio,
+                }
+            } else if aspect_ratio < max_w_max_h {
+                // exits through max.width
+                Size {
+                    width: self.max.width,
+                    height: self.max.width * aspect_ratio,
+                }
+            } else {
+                // exits through max.height
+                Size {
+                    width: self.max.height * aspect_ratio.recip(),
+                    height: self.max.height,
+                }
+            }
+        } else {
+            // final case is where we hit constraints on the min height line
+            if width < self.min.width {
+                // take the point on the min height
+                Size {
+                    width: self.min.height * aspect_ratio.recip(),
+                    height: self.min.height,
+                }
+            } else if aspect_ratio > max_w_max_h {
+                // exit thru max height
+                Size {
+                    width: self.max.height * aspect_ratio.recip(),
+                    height: self.max.height,
+                }
+            } else {
+                // exit thru max width
+                Size {
+                    width: self.max.width,
+                    height: self.max.width * aspect_ratio,
+                }
+            }
+        }
+    }
+
+    /// Sets the max on a given axis to infinity.
+    pub fn unbound_max(&self, axis: Axis) -> Self {
+        match axis {
+            Axis::Horizontal => self.unbound_max_width(),
+            Axis::Vertical => self.unbound_max_height(),
+        }
+    }
+
+    /// Sets max width to infinity.
+    pub fn unbound_max_width(&self) -> Self {
+        let mut max = self.max();
+        max.width = f64::INFINITY;
+        BoxConstraints::new(self.min(), max)
+    }
+
+    /// Sets max height to infinity.
+    pub fn unbound_max_height(&self) -> Self {
+        let mut max = self.max();
+        max.height = f64::INFINITY;
+        BoxConstraints::new(self.min(), max)
+    }
+
+    /// Shrinks the max dimension on the given axis.
+    /// Does NOT shrink beyond min.
+    pub fn shrink_max_to(&self, axis: Axis, dim: f64) -> Self {
+        match axis {
+            Axis::Horizontal => self.shrink_max_width_to(dim),
+            Axis::Vertical => self.shrink_max_height_to(dim),
+        }
+    }
+
+    /// Shrinks the max width to dim.
+    /// Does NOT shrink beyond min width.
+    pub fn shrink_max_width_to(&self, dim: f64) -> Self {
+        let mut max = self.max();
+        max.width = f64::max(dim, self.min().width);
+        BoxConstraints::new(self.min(), max)
+    }
+
+    /// Shrinks the max height to dim.
+    /// Does NOT shrink beyond min height.
+    pub fn shrink_max_height_to(&self, dim: f64) -> Self {
+        let mut max = self.max();
+        max.height = f64::max(dim, self.min().height);
+        BoxConstraints::new(self.min(), max)
```

Comment:
Should there also be something like `f64::min(dim, max.height)` in here, to keep from increasing the max?

Reply:
This (along with `Axis`) is copied without change from Druid. Not that there's no chance of flaws there, but it's not something I'm focusing on at the moment.